knora/dsplib/utils/xml_upload.py

```diff
@@ -910,94 +918,340 @@ def xml_upload(input_file: str, server: str, user: str, password: str, imgdir: s
+            except BaseError:
+                print(f"ERROR while trying to create resource '{resource.label}' ({resource.id}). ")
+                failed_uploads.append(resource.id)
+                continue
 
         permissions_tmp = permissions_lookup.get(resource.permissions)
 
-        try:
-            # create a resource instance (ResourceInstance) from the given resource in the XML (XMLResource)
-            resclass_type = resclass_name_2_type[resource.restype]
-            properties = resource.get_propvals(res_iri_lookup, permissions_lookup)
-            resclass_instance: ResourceInstance = resclass_type(
-                con=con,
-                label=resource.label,
-                iri=resource_iri,
-                permissions=permissions_tmp,
-                bitstream=resource_bitstream,
-                values=properties
-            )
-            resclass_instance = resclass_instance.create()
-        except BaseError as err:
-            print(f"ERROR while trying to create resource '{resource.label}' ({resource.id}). "
-                  f"The error message was: {err.message}")
-            failed_uploads.append(resource.id)
-            continue
-        except Exception as exception:
-            print(f"EXCEPTION while trying to create resource '{resource.label}' ({resource.id}). "
-                  f"The exception message was: {exception}")
+        # create the resource in knora
+        resclass_instance: ResourceInstance = None
+        for _ in range(5):
+            try:
+                resclass_type = resclass_name_2_type[resource.restype]
+                properties = resource.get_propvals(id2iri_mapping, permissions_lookup)
+                resclass_instance = resclass_type(
+                    con=con,
+                    label=resource.label,
+                    iri=resource_iri,
+                    permissions=permissions_tmp,
+                    bitstream=resource_bitstream,
+                    values=properties
+                )
+                resclass_instance = resclass_instance.create()
+                break
+            except ConnectionError:
+                print(f'{datetime.now().isoformat()}: Try reconnecting to DSP server...')
+                time.sleep(1)
+                continue
+            except RequestException:
+                print(f'{datetime.now().isoformat()}: Try reconnecting to DSP server...')
+                time.sleep(1)
+                continue
+            except BaseError:
+                break
+
+        if not resclass_instance or not resclass_instance.iri:
+            print(f"ERROR while trying to create resource '{resource.label}' ({resource.id}). ")
             failed_uploads.append(resource.id)
-            continue
+        else:
+            id2iri_mapping[resource.id] = resclass_instance.iri
+            print(f"Created resource '{resclass_instance.label}' ({resource.id}) with IRI '{resclass_instance.iri}'")
```

Comment:
Should this always be printed or only in case of verbose=True? It could be very verbose in case of hundreds or thousands of resources.

Reply:
This results in 1 line per resource: either success or error. This was present since the early days, and is actually very helpful to see how the xmlupload makes progress.